Fix EdgeReadIterator

michaelgatzen opened the pull request:
- Fixing #1615 
  - EdgeReadIterator was calling AbstractLocusIterator's `getCurrentInterval()` to determine the read overlaps. However, the `lastInterval` that it uses internally was never updated. This PR removes the method `getCurrentInterval()` and instead implements this functionality in `EdgeReadIterator`, which is the only place this has ever been called from (at least in Picard and GATK). **This is arguably a breaking API change**, although I would argue that this method has never worked and should not be placed in AbstractLocusIterator, but instead in an iterator implementation that actually does keep track of the intervals.
  - Keeping track of the intervals has now been moved to AbstractLocusIterator
  - Added a new function that determines if a read is fully contained by the current interval, in that case we don't need to determine the overlaps
  - Renamed a few variables to better describe their use
- Added tests to catch previously faulty behavior
  - Also extended existing tests with the possibility to define custom headers
- Fixed another bug in AbstractLocusIterator and IntervalListReferenceSequenceMask, where the documentation says that it would sort an unsorted input IntervalList, but it never does

michaelgatzen committed:
Fixing bug in EdgeReadIterator

- Adding tests to catch previously faulty behavior
- Also fixing other bugs regarding sorting of IntervalLists in AbstractLocusIterator and IntervalListReferenceSequenceMask

lbergelson commented on src/main/java/htsjdk/samtools/util/EdgeReadIterator.java:
```diff
@@ -88,73 +139,86 @@ public EdgeReadIterator(final SamReader samReader, final IntervalList intervalLi
-             * We put two records to the accumulator. The first one has the "begin" type which corresponds to the locus 
-             * where the block starts. The second one has the "end" type which corresponds to the other locus where the block ends. 
-            */
+            // Let's assume an alignment block starts in some locus.
```
This is all pretty confusing.  It could really use a nice ASCII art if you're feeling creative.

michaelgatzen replied:
```\o/```

lbergelson commented on src/main/java/htsjdk/samtools/util/IntervalListReferenceSequenceMask.java:
```diff
@@ -46,10 +46,9 @@ public class IntervalListReferenceSequenceMask implements ReferenceSequenceMask
-            intervalList.sorted();
-        }
-        final List<Interval> uniqueIntervals = intervalList.uniqued().getIntervals();
+        final boolean isSorted = intervalList.getHeader().getSortOrder() == SAMFileHeader.SortOrder.coordinate;
```
good catch here

lbergelson commented on src/test/java/htsjdk/samtools/util/EdgeReadIteratorTest.java:
```diff
@@ -398,14 +402,124 @@ public void testIntersectingIntervalWithComplicatedCigar() {
+        };
+
+        EdgeReadIterator iterator = new EdgeReadIterator(builder.getSamReader(), intervals);
+        iterator.iterator(); // We need this call to initialize the samReader. Usually a call to hasNext() would do that, which we don't need here.
```
This is a bug here... next() should be valid without calling iterator() or hasNext().  Maybe we should fix that in AbstractLocusIterator.  I think there's a different bug where calling next() returns null instead of throwing NoSuchElementException.

michaelgatzen replied:
Fixed in AbstractLocusIterator

lbergelson reviewed (changes requested):
@michaelgatzen Thanks for taking on this unenviable task.  This set of of classes is quite confusing.  I think I found a subtle (and in practice probably not important) bug with using the IntervalComparator if the intervals happen to have names or strands defined.  I have a few stylistic comments too.  I agree the breaking change is fine since it's so broken to begin with.

michaelgatzen commented on src/main/java/htsjdk/samtools/util/AbstractLocusIterator.java:
```diff
@@ -176,8 +171,14 @@ public AbstractLocusIterator(final SamReader samReader, final IntervalList inter
+                throw new SAMException("The sequence dictionary of the interval list file differs from the sequence dictionary of the input SAM file.");
+            }
+
+            final boolean isSorted = intervalList.getHeader().getSortOrder() == SAMFileHeader.SortOrder.coordinate;
```
Yes good question. First of all, yes, we need both the SAM file and the interval list file to be coordinate sorted. AbstractLocusIterator logs a warning if the sort order is unspecified, and throws an exception if it is specified but not coordinate. What I'm doing here is checking if the interval list is coordinate sorted, because if it is, we might as well skip the sorting. Or do you think we should sort either way?

michaelgatzen commented on src/main/java/htsjdk/samtools/util/EdgeReadIterator.java:
```diff
@@ -69,13 +75,58 @@ public EdgeReadIterator(final SamReader samReader, final IntervalList intervalLi
+         }
+        // Here we need to update the currentInterval. We have to do this using an
+        // IntervalCoordinateComparator to take factor in the order in the sequence dictionary.
+        while (nextInterval != null && intervalCoordinateComparator.compare(new Interval(rec), nextInterval) > 0) {
```
Another good point (and good catch!). I'm using the IntervalCoordinateComparator mostly for convenience, because it takes into account the sort order of the contigs in the sequence dictionary, because they might not be alphabetically sorted. And you're right, it could happen that having a strand or name specified could affect the comparison, but that happens only after looking at start and end coordinates. So this would only be an issue if the current read happens to have the same start and end coordinates as the interval (which would be very rare), and in such cases I think we can afford that we get this wrong, because worst case we will process a few reads with the slower implementation.

michaelgatzen committed:
Addressed PR comments

michaelgatzen commented on src/main/java/htsjdk/samtools/util/EdgeReadIterator.java:
```diff
@@ -69,13 +75,58 @@ public EdgeReadIterator(final SamReader samReader, final IntervalList intervalLi
+     * @param rec The record we want to consider
+     * @return True, if rec is fully contained in the current interval, otherwise false
+     */
+     protected boolean intervalCompletelyContainsRead(final SAMRecord rec) {
```
I can't decide what's better. Clearly for readability it'd be better to advance the iterator and then check if it contains the read, but if I have it in one function, the calling line is quite elegant:
```
final boolean needToConsiderIntervals = intervals != null && !advanceCurrentIntervalAndCheckIfIntervalContainsRead(rec);
```

Here, the function is only called if `intervals != null` and the assignment can be done to a final variable. Or can you think of another way to accomplish this, without unnecessary function calls (for when `intervals == null`) or not having a final variable?

michaelgatzen commented on src/main/java/htsjdk/samtools/util/EdgeReadIterator.java:
```diff
@@ -69,13 +75,58 @@ public EdgeReadIterator(final SamReader samReader, final IntervalList intervalLi
     public EdgeReadIterator(final SamReader samReader, final IntervalList intervalList, final boolean useIndex) {
-        super(samReader, intervalList, useIndex);
+        super(samReader,
+                intervalList == null || intervalList.getHeader().getSortOrder() == SAMFileHeader.SortOrder.coordinate ? intervalList : intervalList.sorted(),
```
Done

michaelgatzen commented on src/main/java/htsjdk/samtools/util/EdgeReadIterator.java:
```diff
@@ -69,13 +75,58 @@ public EdgeReadIterator(final SamReader samReader, final IntervalList intervalLi
+        } else {
+            // For the easy case of a read being fully contained in an interval, we use an iterator to keep track
+            // of the current interval.
+            intervalListIterator = getIntervals().iterator();
```
Done

michaelgatzen commented:
@lbergelson Thanks for the review, and all the deep digging you did for that! I addressed most of the comments, feel free to comment or resolve the threads!

lbergelson commented on src/main/java/htsjdk/samtools/util/EdgeReadIterator.java:
```diff
@@ -142,10 +133,33 @@ void accumulateSamRecord(SAMRecord rec) {
+            // Example: Read (or more accurately, AlignmentBlock) from position 101 to 108
+            //
+            // accumulator (ArrayList<LocusInfo>)     30  31  32  33  34  35  36  37  38 (has one LocusInfo at each position, corresponding to the genomic position)
+            // LocusInfo objects (with genomic pos.) 100 101 102 103 104 105 106 107 108 (the LocusInfo objects can contain EdgingRecordAndOffset objects, if a record starts or ends there)
```
Thanks for this.   I think it's helpful for future viewers.

lbergelson reviewed (approved):
@michaelgatzen I think this is fine to merge.  If we encounter problems with too strict sequence dictionary checking we can revisit it.
